### DISCUSSION TITLE
Fastapi swagger docs with GCP API gateway

### DISCUSSION THREAD

**Discussion by samshipengs** [fastapi/fastapi#9225] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI


app = FastAPI(
    title="My API",
    version="0.0.0",
)

"""
# openapi2-run.yaml
swagger: "2.0"
info:
  title: my-api-id
  description: Sample API on API Gateway with a Cloud Run backend
  version: 1.0.0
schemes:
  - https
produces:
  - application/json
x-google-backend:
  address: my-cloud-run-url
paths:
  /docs:
    get:
      summary: Swagger page.
      operationId: docs
      responses:
        "200":
          description: A successful response
          schema:
            type: string
  /hello:
...
"""
```

### Description

I'm using the GCP API gateway to route API calls to my endpoints written with FastAPI on Cloud Run.

I added API key requirement to the https calls into the API gateway. I can hit other get/post endpoints on cloud run fine, however, I'd like to be able to see and interact with the Swagger docs/ page, I tried adding it in the `openapi2-run.yaml` (see above), but on the browser, it gives me this error:

```
{"code":404,"message":"The current request is not defined by this API."}
```
Does anyone know what the issue might be?

(I'm fairly new to fastapi and gcp API gateway, please let me know if there is misunderstanding)

### Operating System

Linux

### FastAPI Version

0.92.0

### Python Version

3.9.16

**Comment by Ryandaydev** [comment]
In order for the Swagger UI at the _/docs_ endpoint to work with the Google API gateway in front of your Google Cloud Run FastAPI app, I have found that you need to define two special paths in the openapi2-functions.yaml used to configure the gateway:
1. /docs/
2. /openapi.json

Here is the top of a working config file that I tested:

```
# openapi2-functions.yaml
swagger: '2.0'
info:
  title: gcp-api-1
  description: GCP API gateway for my API
  version: 1.0.0
schemes:
  - https
produces:
  - application/json
paths:
  /docs:
    get:
      summary: Docs endpoint
      operationId: api_docs
      x-google-backend:
        address: https://someurl.run.app/docs/
      responses:
        '200':
          description: Docs page
          schema:
            type: string
  /openapi.json:
    get:
      summary: OAS file
      operationId: oas_file
      x-google-backend:
        address: https://someurl.run.app/openapi.json
      responses:
        '200':
          description: OAS file
          schema: 
            type: string
(the rest of the file with a reference to routes you want to call)
```

I'm assuming that you have followed the necessary steps described in this quickstart:
https://cloud.google.com/api-gateway/docs/secure-traffic-gcloud

**Reply by Ryandaydev** [reply]
@Kludex I think this is the answer.

**Reply by Ryandaydev** [reply]
@samshipengs does this answer your question?

**Reply by Celudor** [reply]
I have another issue with /docs and API Gateway. When I add API key for securing gateway /openapi.json is called without api key.

```
 securityDefinitions:
    api_key:
      type: "apiKey"
      name: "key"
      in: "query"
```
![image](https://github.com/user-attachments/assets/9f3eb0fa-1131-4a68-80e0-d81ec5727d9e)

Do you know how to solve this issue?
